Draft: Work in progress adding filter to customize Default DurationOptions.

ajskelton opened the pull request:
The filter works, but something in the DateTimeRange useEffect is reverting back to the default values.

<!--
Please do your best to fill out this template.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code ideally includes documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability

### Credits
<!-- Please list any and all contributors on this PR so that they can be properly credited within this project. -->
Props @username, @username2, ...

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/GatherPress/gatherpress/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

mauteri commented on src/helpers/datetime.js:
```diff
@@ -56,43 +57,6 @@ export const defaultDateTimeEnd = moment
- * @property {string}         label - The human-readable label for the duration option.
- * @property {number|boolean} value - The value representing the duration in hours, or `false` if a custom end time is to be set.
- */
-export const durationOptions = [
```
let's keep the name `durationOptions`, but convert this to a function. Then apply your filter to the function before returning.

mauteri commented on src/components/Duration.js:
```diff
@@ -35,11 +36,11 @@ const Duration = () => {
 			label={__('Duration', 'gatherpress')}
 			value={
-				durationOptions.some((option) => option.value === duration)
+				DurationOptions.some((option) => option.value === duration)
```
this will become `durationOpions().some...` after changes below.

mauteri commented on src/components/Duration.js:
```diff
@@ -35,11 +36,11 @@ const Duration = () => {
 					: false
 			}
-			options={durationOptions}
+			options={DurationOptions}
```
`options={durationOptions()}`

mauteri reviewed (changes requested):
Thanks for contributing, Anthony! Take a look at my comments and let me know what you think.

ajskelton commented:
The comments made a lot of sense and I pushed up a new commit with the updates. With those in place the filter from a theme was working great, but there is one bug that I was able to drill down.

When the page loads the `gatherpress/datetime` store builds the DEFAULT_STATE. The duration value uses the `getDateTimeOffset()` function that calculates what the duration is. During this first instance the durationOptions() function does not use the filtered values from the theme, it still uses the defaults. When it doesn't find the corresponding value in the object returned from durationOptions it returns false. With the state set to false it doesn't render the dropdown and instead shows the Date and Time End field so you can't change the duration anymore. 

Do you think it would be better to save the duration value to the post meta along with the start and end dates so this calculation doesn't need to happen?

mauteri commented:
> The comments made a lot of sense and I pushed up a new commit with the updates. With those in place the filter from a theme was working great, but there is one bug that I was able to drill down.
> 
> When the page loads the `gatherpress/datetime` store builds the DEFAULT_STATE. The duration value uses the `getDateTimeOffset()` function that calculates what the duration is. During this first instance the durationOptions() function does not use the filtered values from the theme, it still uses the defaults. When it doesn't find the corresponding value in the object returned from durationOptions it returns false. With the state set to false it doesn't render the dropdown and instead shows the Date and Time End field so you can't change the duration anymore.
> 
> Do you think it would be better to save the duration value to the post meta along with the start and end dates so this calculation doesn't need to happen?

Hey @ajskelton, can you try to call this function to get the duration value? This should serve the same purpose you are looking for with storing meta. Basically, it takes the start/end datetime and returns the duration or false (same as you'd get from the meta) https://github.com/GatherPress/gatherpress/blob/main/src/helpers/datetime.js#L127 Let me know if that works. Thx!

ajskelton commented:
@mauteri This is the function that doesn't use the filtered values when the default state is being set.

I tested by adding some console.logs in the .find(). When you refresh the page and select the block you see the default duration values logging to the console instead of the filtered duration. 

So it sets the duration to false before checking against the filtered options. 

When it runs here: https://github.com/GatherPress/gatherpress/blob/4b894749193ca340d3cea082e1cac8ef5c6a0e4f/src/stores/datetime.js#L23, it must be running before the filter in the theme is registered.

mauteri commented:
> @mauteri This is the function that doesn't use the filtered values when the default state is being set.
> 
> I tested by adding some console.logs in the .find(). When you refresh the page and select the block you see the default duration values logging to the console instead of the filtered duration.
> 
> So it sets the duration to false before checking against the filtered options.
> 
> When it runs here:
> 
> https://github.com/GatherPress/gatherpress/blob/4b894749193ca340d3cea082e1cac8ef5c6a0e4f/src/stores/datetime.js#L23
> 
> , it must be running before the filter in the theme is registered.

Oh gotcha, let me give it some thought.

carstingaxion commented:
Hello @ajskelton & @mauteri,

are you aware of https://github.com/WordPress/gutenberg/issues/60971, which seems slightly related?

ajskelton commented:
@mauteri I found a fix for the issue. Instead of running the getDateTimeOffset() function in the DEFAULT_STATE, i changed that initial duration value to be null. Then I moved that function to the getDuration selector. When the Duration Component renders it always runs the getDuration selector to dynamically get the value based on the start and end times. And when this runs all of the theme filters have already been registered. 

We do need to check if the current state is false here in the selector. I first tried just using the getDateTimeOffset but when the value was set to false the state wasn't saving the false value and the DateTimeEnd was never rendering. 

I don't have a ton of experience working with custom stores so let me know if this is the wrong way to go about this.

@carstingaxion That new Date/Time Range component does look nice. I think it's still just design right now, I couldn't see any code commits to check out.

mauteri commented on src/stores/datetime.js:
```diff
@@ -78,7 +78,7 @@ const store = createReduxStore('gatherpress/datetime', {
 		getDateTimeStart: (state) => state.dateTimeStart,
 		getDateTimeEnd: (state) => state.dateTimeEnd,
-		getDuration: (state) => state.duration,
+		getDuration: (state) => state.duration === false ? false : getDateTimeOffset(),
```
```
getDuration: (state) => false === state.duration ? false : getDateTimeOffset(),
```

mauteri commented:
This can be tested with this gist. https://gist.github.com/mauteri/5b8402d29de94502fff1e0da9e68f2d8
